oci/pull.bzl

```diff
@@ -0,0 +1,99 @@
+"Pull image layers using Bazel downloader"
+
+_attrs = {
+    "image": attr.string(),
+    "tag": attr.string(),
+}
+
+def _download(rctx, tag, output, type = "manifests"):
+    if type != "blobs" and type != "manifests":
+        fail("type must be blobs or manifests")
+    if rctx.attr.image.startswith("gcr.io/"):
+        registry_url = "https://gcr.io/v2/{image}/{type}/{tag}".format(
+            image = rctx.attr.image[len("gcr.io/"):],
+            type = type,
+            tag = tag,
+        )
+    else:
+        fail("Unrecognized registry. We only understand gcr.io currently. File an issue on rules_oci.")
+
+    sha256 = None
+    if tag.startswith("sha256:"):
+        sha256 = tag[len("sha256:"):]
+    rctx.download(
+        url = registry_url,
+        sha256 = sha256,
+        output = output,
+    )
+    if type == "manifests":
+        return json.decode(rctx.read(output))
+    return None
+
+def _download_blobs(rctx, tag, output):
+    _download(rctx, tag, output, "blobs")
+
+_build_file_header = """\
+# Generated by oci_pull
+
+load("@contrib_rules_oci//oci:defs.bzl", "oci_image")
+
+package(default_visibility = ["//visibility:public"])
+"""
+
+_oci_image_target = """\
+oci_image(
+    name = "{name}_{os}_{arch}",
+    tars = {tars},
+    os = "{os}",
+    architecture = "{arch}",
+)
+"""
+
+_alias_target = """\
+alias(
+    name = "{name}",
+    actual = select({{
+        "@platforms//cpu:arm64": "{name}_linux_arm64",
+        "@platforms//cpu:x86_64": "{name}_linux_amd64",
+    }}),
+)
+"""
+
+def _pull_impl(rctx):
+    if not rctx.attr.tag.startswith("sha256:"):
+        # buildifier: disable=print
+        print("""\
+WARNING: unpinned tag {} may change in future installs.
+To get the digest, you can run
+$ bazel run @oci_crane_darwin_arm64//:crane digest gcr.io/distroless/static:latest
+Then replace the tag with the sha256:... value it prints.
+""".format(rctx.attr.tag))
+
+    build_content = [_build_file_header]
+
+    mf = _download(rctx, rctx.attr.tag, rctx.attr.image + ".json")
+
+    for m in mf["manifests"]:
+        image_mf_file = "{}_{}.json".format(m["platform"]["os"], m["platform"]["architecture"])
+        image_mf = _download(rctx, m["digest"], image_mf_file)
+
+        tars = []
+        for layer in image_mf["layers"]:
+            sha256 = layer["digest"].replace("sha256:", "")
+            _download_blobs(rctx, layer["digest"], sha256 + ".tar")
+            tars.append(sha256 + ".tar")
+
+        build_content.append(_oci_image_target.format(
```

Comment:
all of these are true, I think we'd want to create different external repo for each platform, maybe using toolchains to select the right one?

Since that would require a macro wrapping several repository rule calls, I think we'd have to tell users to check in the manifest-of-images (the JSON with the manifest for each platform) so that we can `json.decode` that in the macro and know all the repository rules we need to call.

Reply:
not sure about the part "using the right one". we could worry about it later. 

> I think we'd have to tell users to check in the manifest-of-images

right. we'd need something like. it could even be a simple dict on the macro. we don't need to know about digests of blobs. a simple `{"os/arch": "digest/tag"}` dict would suffice.

Reply:
Take another look - I'm using a mirror script to pull the manifest lists into starlark so we can operate over them with a macro.

Reply:
new approach, matching the `mediaType` against 
```
_MANIFEST_TYPE = "application/vnd.docker.distribution.manifest.v2+json"
_MANIFEST_LIST_TYPE = "application/vnd.docker.distribution.manifest.list.v2+json"
```
to handle the two cases correctly